crates/python/src/api.rs

```diff
@@ -102,6 +102,15 @@ py_wrap_error!(
     PyRuntimeError
 );
 
+/// Get the keyword `key` value from `kwds` if it is of type `Option<T>`, else `None`.
+/// Note that values present at `key` that cannot be extracted into `Option<T>` are treated as `None`.
+fn get_kwd<'a, T: FromPyObject<'a>>(kwds: Option<&'a PyDict>, key: &str) -> Option<T> {
+    kwds
+        .and_then(|kwds| kwds.get_item(key))
+        .map(|value| value.extract())
```

Comment:
Need to double check this, but clippy may point this out:

```suggestion
        .map(PyAny::extract)
```

Reply:
I added this locally at `<T>::extract`

Reply:
Actually, I'm not sure how pyo3 handles this, but I want to extract it into `Option<T>` so I can handle the case of a key being present, but the value being `None`,

```rs
fn get_kwd<'a, T: FromPyObject<'a>>(kwds: Option<&'a PyDict>, key: &str) -> PyResult<Option<T>> {
    kwds.and_then(|kwds| kwds.get_item(key))
        .map(<Option<T>>::extract)
        .unwrap_or(Ok(None))
}
```

Reply:
`pyo3` should handle it just fine if you tell it to extract an `Option`. I'd suggest writing it like this:

```rust
fn get_kwd<'a, T: FromPyObject<'a>>(kwds: Option<&'a PyDict>, key: &str) -> PyResult<Option<T>> {
    kwds.and_then(|kwds| kwds.get_item(key))
        .map(PyAny::extract::<Option<T>>)
        .unwrap_or(Ok(None))
}
```

The method is `PyAny::extract<D>`, so you want to specify the type to extract to _after_ the method name.

FWIW, the handling of a `None` dictionary can also be written like this (using my above suggestion as a base):

```rust
fn get_kwd<'a, T: FromPyObject<'a>>(kwds: Option<&'a PyDict>, key: &str) -> PyResult<Option<T>> {
    kwds.and_then(|kwds| kwds.get_item(key))
        .map_or(Ok(None), PyAny::extract::<Option<T>>)
}
```

Reply:
On second read, does "so I can handle the case of a key being present, but the value being None" mean that "key not present" should be a different return value from "key present, `value = None`"? I think you'd need `Option<Option<T>`` at that point...

Reply:
No need to distinct the cases.

Welp, I wrote these tests before learning about [this issue](https://pyo3.rs/v0.13.1/faq.html#i-cant-run-cargo-test-im-having-linker-issues-like-symbol-not-found-or-undefined-reference-to-_pyexc_systemerror) but these are my expecations:

```rs
#[cfg(test)]
mod tests {
    use super::get_kwd;
    use pyo3::{self, types::PyDict, PyResult, Python};

    #[test]
    fn test_get_kwd() {
        pyo3::prepare_freethreaded_python();
        let result: PyResult<()> = Python::with_gil(|py| {
            assert!(
                matches!(get_kwd::<bool>(None, "key"), Ok(None)),
                "kwds `None` should successfully return empty value."
            );

            assert!(
                matches!(get_kwd::<bool>(Some(PyDict::new(py)), "key"), Ok(None)),
                "empty `kwds` should successfully return empty value."
            );

            assert!(
                matches!(
                    get_kwd::<bool>(
                        Some({
                            let kwds = PyDict::new(py);
                            kwds.set_item("key", None::<Option<bool>>)?;
                            kwds
                        }),
                        "key"
                    ),
                    Ok(None)
                ),
                "`kwds` with key but empty value should successfully return empty value."
            );

            assert!(
                matches!(
                    get_kwd::<bool>(
                        Some({
                            let kwds = PyDict::new(py);
                            kwds.set_item("key", true)?;
                            kwds
                        }),
                        "key"
                    ),
                    Ok(Some(true))
                ),
                "`kwds` with key and value of matching type should successfully return that value."
            );

            assert!(
                matches!(
                    get_kwd::<bool>(
                        Some({
                            let kwds = PyDict::new(py);
                            kwds.set_item("key", 4u64)?;
                            kwds
                        }),
                        "key"
                    ),
                    Err(_)
                ),
                "`kwds` with a value of disagreeing type should return an error."
            );

            Ok(())
        });

        result.expect("get_kwd test should pass without issue.");
    }
}

```